source/common/grpc/buffered_async_client.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Grpc {
+
+enum class BufferState { Buffered, PendingFlush };
+
+template <class RequestType, class ResponseType> class BufferedAsyncClient {
+public:
+  BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
+                      Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+      : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
+        client_(client) {}
+
+  virtual ~BufferedAsyncClient() { cleanup(); }
+
+  uint32_t publishId(RequestType& message) { return MessageUtil::hash(message); }
```

Comment:
Do we want to use `MessageUtil::hash` as ID? it is possible that we have two identical log entries if we omit time/request IDs?

Reply:
I think we don't have to consider "log" in this class, and the method publishing ID should be pluggable. Should we provide this as interface?  With this, we don't have to take care about what identifiers provided messages have.

Reply:
Or just a sequential per client because those ID are thread local / stream local?

Reply:
Stream local identifier also good to me.